### PR TITLE
✏️ Feat/ #22 Application Update methods

### PR DESCRIPTION
**Pull request by demariadaniel**
Feat: #22: Application Update Methods

## Summary

- Adds Service Method to update Application details, with related API route

### Related Issues

- [22](https://github.com/orgs/Pan-Canadian-Genome-Library/projects/1?pane=issue&itemId=86248018&issue=Pan-Canadian-Genome-Library%7Cdaco%7C22)

## Description of Changes
- Adds Application Service `editApplication` method for updating application contents
- Method sets application state to DRAFT
- Adds API `editApplication` function which validates application state allows edits
- Adds first Express route for Applications
- Updates to test files, and drizzle config / dbml config related to PR discussion on: 
https://github.com/Pan-Canadian-Genome-Library/daco/pull/40/files

### Special Instructions
Changes to Drizzle config require running `pnpm run build` to allow `drizzle-kit` operations
`drizzle-kit` now uses `/dist` for schema lookup

## Readiness Checklist

- [x] **Self Review**
  - I have performed a self review of code
  - I have run the application locally and manually tested the feature
- [x] **PR Format**
  - The PR title is properly formatted to match the pattern: `#{TicketNumber}: Description of Changes`
  - Links are included to all relevant tickets
- [x] **Labels Added**
  - Label is added for each package/app that is modified (`api`, `ui`, `data-model`, etc.)
  - Label is added for the type of work done in this PR (`feature`, `fix`, `chore`, `documentation`)
- [x] **Local Testing**
  - Successfully built all packages locally
  - Successfully ran all test suites, all unit and integration tests pass
- [x] **Updated Tests**
  - Unit and integration tests have been added that describe the bug that was fixed or the features that were added
- [x] Documentation
  - All new environment variables added to `.env.schema` file and documented in the README
  - All changes to server HTTP endpoints have open-api documentation
  - All new functions exported from their module have TSDoc comment documentation

### CODE REVIEW

File: apps/api/package.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
 	"scripts": {
 		"build": "rimraf ./dist && tsc",
 		"dev": "tsx watch --env-file=.env ./src/main.ts",
-		"test": "npx tsx --test --test-reporter=spec --experimental-test-coverage ./src/tests/**/*",
+		"test": "tsx --test --test-reporter=spec --experimental-test-coverage ./tests/**/*.test.ts",
```
Comment (demariadaniel) [inline apps/api/package.json:9]: Feedback from https://github.com/Pan-Canadian-Genome-Library/daco/pull/40

File: apps/api/tests/api/index.test.ts [hunk 2]
```diff
@@ -20,7 +20,7 @@
 import assert from 'node:assert';
 import { after, describe, it } from 'node:test';
 
-import { port } from '../server.js';
+import { port } from '../../src/server.js';
```
Comment (demariadaniel) [inline apps/api/tests/api/index.test.ts:23]: `test` folder moved out of `src`

**Review by JamesTLopez — COMMENTED** [review]
Minor stuff and a suggestion, otherwise lgtm 🔥

**Comment by demariadaniel** [comment]
Endpoint tested with Postman:
Screenshot 2024-12-02 at 2 43 31 PM

File: apps/api/src/routes/application-router.ts [hunk 3]
```diff
@@ -23,15 +23,15 @@ import { editApplication } from '../api/application-api.js';
...
+	const result = await editApplication({ id, update });
 
-	if (record) {
-		res.send(record);
+	if (result.success) {
+		res.send(result.data);
 	} else {
-		res.status(500).send({ error: `Error updating application with id ${id}` });
+		res.status(500).send({ message: result.message, errors: result.errors });
 	}
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply apps/api/src/routes/application-router.ts]: Yes fair, the `else` could be removed. Generally we prefer putting the 'truthy' condition first but your recommendation is still applicable.

Also in the code snippet above I think you copied the Git diff, double res.send's would trigger an error. 
Just to be sure I understand you right, you're saying `else` is not needed here.

[the remark this answers is not in the document]
Comment (demariadaniel) [reply apps/api/src/routes/application-router.ts]: edit: My editor seems to prefer keeping the `else` as a type guard to enforce `result is failure`

[the remark this answers is not in the document]
Comment (JamesTLopez) [reply apps/api/src/routes/application-router.ts]: Ah yes my bad, there would be a return statement after the correct `res`. I have updated it to what my initial thoughts were.

File: apps/api/README.md [hunk 4]
```diff
@@ -25,3 +25,7 @@ This app uses features from Node ^20.9, in combination with TSX for TypeScript s
 ### Database
 
 - This application implements a Postgres database based on the following [data model](../../docs/model/README.md) and managed by [Drizzle ORM](https://orm.drizzle.team/docs/overview)
+
+- Drizzle is configured to read the schema from the build directory (`/dist`) due to a known Drizzle Kit issue with ESM imports. Any Drizzle Kit operations (including migrations) require first building the application with `pnpm run build`.
+
+- A schema dbml (database markup language) file can be generated using the script `pnpm run dbml`. This file is found at `./src/db/schema.dbml`.
```
Comment (leoraba) [inline apps/api/README.md:31]: 👍

File: apps/api/tsconfig.json [hunk 5]
```diff
@@ -6,5 +6,5 @@
 		"module": "NodeNext",
 		"moduleResolution": "NodeNext"
 	},
-	"include": ["src", "./drizzle.config.ts"]
+	"include": ["src", "./drizzle.config.ts", "tests"]
```
Comment (leoraba) [inline apps/api/tsconfig.json:9]: is there a reason why we are including `test` folder in here?

File: apps/api/src/routes/application-router.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+ */
+
+import bodyParser from 'body-parser';
+import express from 'express';
+import { editApplication } from '../api/application-api.js';
+
+const applicationRouter = express.Router();
+const jsonParser = bodyParser.json();
+
+applicationRouter.post('/application/edit/', jsonParser, async (req, res) => {
```
Comment (leoraba) [inline apps/api/src/routes/application-router.ts:27]: seems this route hasn't been included in the Swagger documentation

**Review by leoraba — COMMENTED** [review]
Great updates in this PR. I’ve left a few comments but so far looks good to me.

File: apps/api/tsconfig.json (hunk 5 above)
Comment (demariadaniel) [reply apps/api/tsconfig.json:9]: This was an auto update from moving folders around, doesn't need to be included

File: apps/api/src/routes/application-router.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+	// TODO: Add Auth & Zod validation
+	const data = req.body;
+	const { id, update } = data;
+	const result = await editApplication({ id, update });
+
+	if (result.success) {
+		res.send(result.data);
+	} else {
+		res.status(500).send({ message: result.message, errors: String(result.errors) });
```
[the remark this answers is not in the document]
Comment (demariadaniel) [reply apps/api/src/routes/application-router.ts:36]: Yes -- if id is not found we're returning a `500`
I can update to a `4xx`

File: apps/api/src/routes/application-router.ts (hunk 6 above)
Comment (demariadaniel) [reply apps/api/src/routes/application-router.ts:27]: Good call -- we just added swagger yesterday. Adding now.

Comment (demariadaniel) [reply apps/api/src/routes/application-router.ts]: Screenshot 2024-12-04 at 12 08 40 PM

File: apps/api/src/routes/application-router.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (demariadaniel) [reply apps/api/src/routes/application-router.ts]: Added basic updates for 404 handling. We don't have a central error handler setup yet so I've added a TODO as well.

File: apps/api/src/resources/swagger.yaml [hunk 8]
```diff
@@ -34,3 +34,152 @@ paths:
...
+          content:
+            application/json:
+              schema:
+                type: object
+        '200':
+          description: 'Application updated'
+          content:
+            application/json:
+              schema:
+                type: object
```
Comment (leoraba) [inline apps/api/src/resources/swagger.yaml:68]: responses schemas can also be defined here using the structures defined by the `success` and `failure` helpers

**Review by leoraba — APPROVED** [review]
LGTM

File: apps/api/src/resources/swagger.yaml (hunk 8 above)
Comment (demariadaniel) [reply apps/api/src/resources/swagger.yaml:68]: Will add in #58

**Pull request merged**